src/Confluent.SchemaRegistry/Rest/RestService.cs

```diff
@@ -0,0 +1,302 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System.Collections.Generic;
+using System.Net;
+using System.Linq;
+using System.Net.Http;
+using System;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json;
+
+
+namespace Confluent.SchemaRegistry
+{
+    /// <remarks>
+    ///     It may be useful to expose this publically, but this is not
+    ///     required by the Avro serializers, so we will keep this internal 
+    ///     for now to minimize documentation / risk of API change etc.
+    /// </remarks>
+    internal class RestService : IRestService
+    {
+        private static readonly string acceptHeader = string.Join(", ", Versions.PreferredResponseTypes);
+
+        /// <summary>
+        ///     The index of the last client successfully used (or random if none worked).
+        /// </summary>
+        private int lastClientUsed;
+        private object lastClientUsedLock = new object();
+
+        /// <summary>
+        ///     HttpClient instances corresponding to each provided schema registry Uri.
+        /// </summary>
+        private readonly List<HttpClient> clients;
+
+        
+        /// <summary>
+        ///     Initializes a new instance of the RestService class.
+        /// </summary>
+        public RestService(string schemaRegistryUris, int timeoutMs)
+        { 
+            this.clients = schemaRegistryUris
+                .Split(',')
+                .Select(uri => uri.StartsWith("http", StringComparison.Ordinal) ? uri : "http://" + uri) // need http or https - use http if not present.
+                .Select(uri => new HttpClient() { BaseAddress = new Uri(uri, UriKind.Absolute), Timeout = TimeSpan.FromMilliseconds(timeoutMs) })
+                .ToList();
+        }
+
+        #region Base Requests
+
+        private async Task<HttpResponseMessage> ExecuteOnOneInstanceAsync(HttpRequestMessage request)
+        {
+            // There may be many base urls - roll until one is found that works.
+            //
+            // Start with the last client that was used by this method, which only gets set on 
+            // success, so it's probably going to work.
+            //
+            // Otherwise, try every client until a successful call is made (true even under 
+            // concurrent access).
+
+            string aggregatedErrorMessage = null;
+            HttpResponseMessage response = null;
+            bool gotOkAnswer = false;
+            bool firstError = true;
+            
+            int startClientIndex;
+            lock (lastClientUsedLock)
+            {
+                startClientIndex = this.lastClientUsed;
+            }
+
+            for (int i = startClientIndex; i < clients.Count + startClientIndex && !gotOkAnswer; ++i)
```

Comment:
This will just try client startClientIndex..end, it should roll-over and start from 0 until it reaches startClientIndex again.

Reply:
it does? it iterates through clients.Count times and modulo arithmetic is used in the loop.